backend/src/main.rs

```diff
@@ -172,8 +180,89 @@ fn get_balance() -> Amount {
 }
 
 #[update]
-fn answer_task(_id: TaskId, _content: Content) -> AnswerId {
-    0
+fn answer_task(task_id: TaskId, content: Content) -> AnswerId {
+    let caller = caller();
+    STATE.with(|s| {
+        let ledger = s.ledger.borrow();
+
+        // Precondition: caller is a principal on the ledger
+        if !ledger.contains_key(&caller) {
+            ic_cdk::trap(&format!("Principal {} cannot provide an answer as this is not a registered\
+             user on the ledger.", caller));
+        }
+
+    });
+
+
+    STATE.with(|s| {
+        let mut tasks = s.tasks.borrow_mut();
+        let mut answers = s.answers.borrow_mut();
+
+        if let Some(task) = tasks.get_mut(&task_id) {
+
+            // Precondition: there are less than max_answers for taskID
+            if task.answers.len() > MAX_NUMBER_ANSWERS {
+                ic_cdk::trap(&format!(
+                    "Cannot provide an answer to task {} as the maximum number of {} answers is already \
+                reached.",
+                    task_id, MAX_NUMBER_ANSWERS));
+            }
+
+            // Precondition: the caller hasn’t submitted an answer for this task
+            for answer_id in task.answers.iter() {
+                match answers.get(answer_id){
+                    Some(answer) => {
+                        if caller == answer.submitter {
+                            ic_cdk::trap(&format!(
+                                "The principal {} already submitted an answer for the task with ID {}.",
+                                caller, task_id));
+                        }
+                    },
+                    // this is a case which should not occur, but let's catch it just to be sure
+                    None => {
+                        ic_cdk::trap(&format!(
+                            "The answer with ID {} was listed in task with ID {} even though there is no\
+                    such answer recorded.",
+                            answer_id, task_id));
+                    }
+                }
+            }
+
+            // Precondition: the deadline for the task has not expired
+            if task.deadline < time() {
+                ic_cdk::trap(&format!(
+                    "No new solution can be provided as the deadline for the task {} has already expired.",
+                    task_id));
+            }
+
+            // Precondition: the solution's size is less than max_solution_size
+            if content.len() > MAX_SOLUTION_SIZE {
+                ic_cdk::trap(&format!(
+                    "Maximum size of solution is {} but {} was given.",
+                    MAX_SOLUTION_SIZE, content.len()
+                ));
+            }
```

Comment:
Let's move these checks before the one that the caller has submitted an answer for this task. Reason: these are cheaper checks than iterating over the existing answers, so better to start with them and filter out quickly before we do the heavier checks.